python/tests/experimental/core/test_udf_schema.py

```diff
@@ -55,6 +61,42 @@ def square(x: Union[Dict[str, List], pd.DataFrame]) -> Union[List, pd.Series]:
     return x["col1"] * x["col1"] if isinstance(x, (pd.Series, pd.DataFrame)) else [xx * xx for xx in x["col1"]]
 
 
+action_list = []
+
+
+def do_something_important(validator_name, condition_name: str, value: Any, column_id=None):
+    print("Validator: {}\n    Condition name {} failed for value {}".format(validator_name, condition_name, value))
+    action_list.append(value)
+    if column_id:
+        # this list is just to verify that the action was called with the correct column id
+        action_list.append(column_id)
+    return
+
+
+@condition_validator(["col1", "add5"], condition_name="less_than_four", actions=[do_something_important])
+def lt_4(x):
+    return x < 4
+
+
+def test_validator_udf_pandas() -> None:
+    global action_list
+    data = pd.DataFrame({"col1": [1, 3, 7]})
+    schema = udf_schema()
+    why.log(data, schema=schema).view()
+    assert 7 in action_list
+
+
+def test_validator_udf_row_with_id() -> None:
```

Comment:
It would be good to cover all the code paths... row, dataframe, homogeneous dataframe